### ISSUE TITLE
OpenAI timeout

### ISSUE THREAD

**Issue by andreapiso** [stanfordnlp/dspy#207]
OpenAI APIs are quite unstable recently and time out often. Training a prompt requires quite a number of calls which means 99% you will experience a timeout. The current max_time is hard-coded as part of the Backoff decorator and set to 1000 seconds. It would be good to be able to change this value so that the compilation would not take forever.

**Comment by raquette** [comment]
Same problem

**Comment by darinkishore** [comment]
@raquette @andreapiso provide stack trace?

**Comment by andreapiso** [comment]
@darinkishore  what stack trace do you need? The program does not crash, it hangs because openAI does not response until it prints that it's retrying with exponential backoff because api.openai.com did not respond.

**Comment by raquette** [comment]
<img width="1343" alt="image" src="https://github.com/stanfordnlp/dspy/assets/16106611/ee36920e-6664-4cb1-b19a-1228facd335b">
@andreapiso I tested with gpt-3.5-turbo-1106 (new version), it is much better.

**Comment by okhat** [comment]
Ah, I think this issue is (in the short term) less severe now?

**Comment by andreapiso** [comment]
I tested it today with 1106 and the OpenAI API still gets stuck fairly often, which makes dspy take a almost two hours to complete a generation of 8 candidate programs with 24 examples.

**Comment by andreapiso** [comment]
In particular, I am starting to get:

```
Error for example in dev set: 		 HTTP code 502 from API (<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
)
```

I recognise the issue is on the OpenAI side but not being able to change the behaviour and the timeout from dspy makes it quite a bit worse.

**Comment by dasjyotishka** [comment]
It is taking me more than two hours to train a chain-of-thought based question-answer predictor having 400 pairs, and often I am facing timeout problems. It would be good to resolve this issue.

**Comment by okhat** [comment]
Can you help me understand the issue? I run large experiments with OpenAI (the latest turbo model at all times) and, except when they note on https://status.openai.com/ that there are issues, it's always very fast and stable.

I understand sometimes they're unreliable. In that case, I don't see how DSPy can help though.

> max_time is hard-coded as part of the Backoff decorator and set to 1000 seconds

We can definitely set this to less than 1000, but how can that help? Won't this just timeout on the next call still?

Basically if OpenAI is inherently unstable at a given hour, how can the library deal with that? The main thing that comes to mind is make sure you set the number of threads to be low enough depending on your rate limit.

Happy to address this if I'm missing some context @dasjyotishka @andreapiso @raquette @darinkishore

**Comment by okhat** [comment]
This may explain the issue from @dasjyotishka ?

![image](https://github.com/stanfordnlp/dspy/assets/963532/8cded450-8670-4f34-9717-7e65d1a33f92)

**Comment by andreapiso** [comment]
> Can you help me understand the issue? I run large experiments with OpenAI (the latest turbo model at all times) and, except when they note on https://status.openai.com/ that there are issues, it's always very fast and stable.
> 
> I understand sometimes they're unreliable. In that case, I don't see how DSPy can help though.
> 
> > max_time is hard-coded as part of the Backoff decorator and set to 1000 seconds
> 
> We can definitely set this to less than 1000, but how can that help? Won't this just timeout on the next call still?
> 
> Basically if OpenAI is inherently unstable at a given hour, how can the library deal with that? The main thing that comes to mind is make sure you set the number of threads to be low enough depending on your rate limit.
> 
> Happy to address this if I'm missing some context @dasjyotishka @andreapiso @raquette @darinkishore

Not sure if the experience is different depending on the geographic region you are in (I am in  South East Asia), but my experience with OpenAI API has been very different, and it consistently drops about 10% of requests (not an exaggeration) especially when submitting multiple requests in parallel (still being very far from the rate cap). The requests being dropped happen at seemingly random so re-submitting the request most of the time works. This is has been experienced by multiple users, on different wifi networks using different machines (and different clients outside of DSPy, we get the same timeouts when calling OpenAI from javascript). 

So, when optimising a prompt in dspy with something like random search, every 10-15 requests effectively dspy gets stuck for 15 minutes, unless I am there to stop the notebook manually and re-run it since the previous iterations of the optimisations get cached.

**Comment by andreapiso** [comment]
It happens consistently with single thread too, which is actually the worst case scenario, because at least with multiple requests in parallel other requests in the optimisation set can stil proceed while one gets stuck, instead the 15 minute break paralyses the whole process every time.

**Comment by okhat** [comment]
Okay, a couple more questions. Does this happen with the latest gpt-3.5-turbo (at whatever time you faced the issue)? It’s possible OpenAI downgrades older models.

Does your code include generation of many completions at once (like n=10) or something like that?

---

Basically, with a single thread, during the “15 minute” gap, the code is retrying the openai call with exponential backoff. If every repeated call keeps failing, that is completely outside our control. I think during 15 minutes, it basically has to retry around 10 times and fail every time. Is that true? In that case, what behavior do you expect from the library?

**Comment by okhat** [comment]
You may benefit from requesting a rate limit increase by the way. The default rate limit (tokens per minute) is absurdly small.

**Comment by andreapiso** [comment]
I am definitely nowhere near hitting the rate limit, this happens even just sending 20-30 calls over the span of the optimisation process. 

```
Basically, with a single thread, during the “15 minute” gap, the code is retrying the openai call with exponential backoff. If every repeated call keeps failing, that is completely outside our control. I think during 15 minutes, it basically has to retry around 10 times and fail every time. Is that true? In that case, what behavior do you expect from the library?
```

The point is that if this was happening, I would see the "backoff retry" message being printed out right? Instead, nothing gets printed out for an absurd amount of time. The expectation would be to allow users to pass the timeout parameter for dspy to set accordingly? So that if I know my calls should last <x> time, I could set a timeout of <x + buffer>.  

I think failing 10 times in 15 mins with backoff would happen for a 60 seconds timeout, but are you sure the timeout is 60 seconds, and not 10 minutes?  (e.g. see here: https://github.com/openai/openai-python/issues/762 ) the experience right now is that every time a call is dropped everything in DSPy hangs for 10+ minutes with no backoff or anything else being printed.

**Comment by okhat** [comment]
The default rate limit can be triggered by as little as 20 calls. Isn’t it something like 90k tokens per minute?

That said, I understand now your concern. It’s not directly related the 15 minute gap. You basically need a timeout for each individual request, not a smaller timeout for the whole backoff logic.

I think that’s reasonable. We can make sure each individual request times out in 40 seconds (though I would have preferred that OpenAI drops the request properly on their end rather than an arbitrary small timeout). We still want to keep the 1000 second logic as-is. Does this sound right to you?

**Comment by vikrantrathore** [comment]
Facing the same issue on self histed LLM which works when token is set to 1000 but timeout when max_token is 2000.

When using openai api SDK directly with the prompt generated by checking inspect_history it works fine. Something wrong with OpenAI gpt3.py in dspy.

Error is same as above 502 Bad Gateway.
